### PR TITLE
#161440329 Create travel dashboard analytics

### PR DESCRIPTION
**Pull request by janetnim**
#### What does this PR do?
Create API to fetch analytics for travel requests from the Travel Administrator's center

#### Description of Task to be completed?
Functionality of: 
1. Get number of total requests
2. Get number of pending requests
3. Get number of people leaving the center
4. Get number of people visiting the center
5. Get the average travel duration 
6. Get the average travel request lead time

#### How should this be manually tested?

- Clone the repo `https://github.com/andela/travel_tool_back.git` and `https://github.com/andela/travel_tool_front.git`
- Checkout to the branch `ft-travela-dashboard-analytics-161374570`
- Run the following commands to install dependencies and start the server and client
```yarn install```
```make start```
- Log in the application and get the token from cookies and add it to `postman` in authorization header.
- Test the API with params as shown in the screenshot
- To download a CSV file, set `type` in params as`file`

#### Any background context you want to provide?
You have to be a travel administrator to access the dashboard page.

#### What are the relevant pivotal tracker stories?
[161440329](https://www.pivotaltracker.com/story/show/161440329)

#### Screenshots (if appropriate)
screen shot 2018-10-26 at 13 31 49

### CODE REVIEW

File: src/modules/analytics/index.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import express from 'express';
+import AnalyticsController from './travelRequests/analyticsController';
+import middlewares from '../../middlewares';
+
+const { authenticate, RoleValidator } = middlewares;
+const Router = express.Router();
+
+// REMEMBER TO PUT ROLE VALIDATOR
```
Comment (Stylll) [inline src/modules/analytics/index.js]: Hi @janetnim Please remove the commented line.

File: src/modules/analytics/travelRequests/__tests__/travelAnalytics.test.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      .set('authorization', travelAdminToken)
+      .end((err, res) => {
+        expect(res.statusCode).toEqual(200);
+        expect(res.body.success).toEqual(true);
+        if (err) return done(err);
+        done();
+      });
+  });
+
+  it('should return a csv report file', (done) => {
```
Comment (Stylll) [inline src/modules/analytics/travelRequests/__tests__/travelAnalytics.test.js]: For this test case, I think you should also check if the content-type and content disposition of the response header is text/csv and attachment.

File: src/modules/analytics/travelRequests/analyticsController.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  static getResult(count, results) {
+    Object.keys(count).forEach((key) => {
+      results.push({ name: key, value: count[key] });
+      return results;
+    });
+  }
+
+  static async analytics(req, res) {
+    try {
+      const { location } = req.user;
```
Comment (ronaldndirangu) [inline src/modules/analytics/travelRequests/analyticsController.js]: I think you should get the location from the req.query passed from the front end.

File: src/modules/analytics/travelRequests/__tests__/travelAnalytics.test.js (hunk 2 above)
Comment (janetnim) [reply src/modules/analytics/travelRequests/__tests__/travelAnalytics.test.js]: I now tested that thanks

File: src/modules/analytics/travelRequests/analyticsController.js (hunk 3 above)
Comment (janetnim) [reply src/modules/analytics/travelRequests/analyticsController.js]: You can get it either way, front-end or backend. It'll still be the same

Comment (ronaldndirangu) [reply src/modules/analytics/travelRequests/analyticsController.js]: Okay it makes sense.

**Pull request merged**